doc/C/preformatted/flvdumper.1.in
.\" -*- coding: us-ascii -*-
.if \n(.g .ds T< \\FC
.if \n(.g .ds T> \\F[\n[.fam]]
.de URL
\\$2 \(la\\$1\(ra\\$3
..
.if \n(.g .mso www.tmac
.TH flvdumper 1 "17 January 2012" "" ""
.SH NAME
flvdumper \- Gnash FLash Video File Dumper
.SH Synopsis
.nf

flvdumper (options)... (file)...
.fi
.SH DESCRIPTION
Dump information about .flv files.
.PP
When a SWF player, including Gnash, plays a SWF "movie",
the movie can contain video files.
Flvdumper prints these files,
which are comprised of an H.263 variant or H.254 video bit stream.
.PP
More information about Flash Video is available at the Wikipedia Flash
Video article: http://en.wikipedia.org/wiki/Flash_Video
.TP 
-h
Print usage info.
.TP 
-m
Print only meta tags (default).
.TP 
-a
Print all the tags.


NEWS
Gnash 0.8.10
YYYY/MM/DD

Improvements since 0.8.9 release are:

 * Qt4 GUI supports mouse wheel, clipboard, and screen resolution.
 * Enhanced UI support for script limits (abort popups, user prefs).
 * BitmapData functions copyPixels(), copyChannel(), perlinNoise() and noise() 
   implemented.
 * Node id mapping in ActionScript XML class implemented (XML.idMap).
 * Fix dispatching of Sound.onLoad event, fixing google dict audio.
 * Fix support for control tags found after last expected frame (#33176).
 * Fix support for uncompressed sound with gstreamer media handler.
 * Implement Button.getDepth(), fix button key events.
 * Fixes to startDrag and stopDrag opcodes.
 * Implement onSoundComplete() for event sounds (#23020).
 * Fix MovieClip.onLoad event dispatching and constant pools handling, fixing
   support for movies generated by the evil Adobe Captivate tool (#33521).
 * Fix unattached Sound.stop() semantic (#33888) enjoy Super Mario!
 * OpenVG renderer added.
 * Improved framebuffer GUI and touchscreen support.
 * Framebuffer now supports using multiple renderers.
 * Refactored input device support.
 * Fix parsing of lossless 16bit bitmaps, fixing support for movies
   generated by the evil TechSmit Camtasia tool (#34625).
 * Improved double buffering support for AGG in a framebuffer.

Gnash 0.8.9
2011/03/19

Caveats:

 * The minimum required FFmpeg version, for people who prefer it over
   Gstreamer, is now 51.28.0.
 * The configure script option --enable-i810-lod-bias, which provided
   a workaround for a very old driver for the Intel I810 when using the
   OpenGL renderer, has been removed.
 * Gnash now uses the system-installed libltdl rather than its in-tree
   copy.
 * NPAPI headers are now required for building the Mozilla plugin; they
   are normally part of the Xulrunner development package.

Improvements since 0.8.8 release are:

 * Fixed playback of uncommon MPEG4 streams when using GStreamer.
 * Avoid drawing operations when the window is obscured, including when
   Gnash is used as a plugin, starting with Firefox 4.0 (GTK GUI).
 * Support SWF with embedded ScreenVideo when using FFMPEG or
   gstreamer-ffmpeg >= 0.10.12.
 * Fix playback of SWF-embedded ADPCM audio when using Gstreamer.
 * Fix playback of SWF-embedded PCM audio when using FFMPEG.
 * Fix swapped _framesloaded and _totalframes properties.
 * Fixes for several crashes, with better automated testing of the
   entire ActionScript API for input sanitization.
 * Improved ActionScript performance processing, especially
   for interactive applications.
 * NPAPI plugin: fix relative urls resolution in presence of <base> tag
   and "base" parameter; improved robustness and compatibility with
   chromium browser. Avoid blocking the parent process.
 * Improved SWF8 support with BitmapData.draw, BitmapData.loadBitmap,
   and other bitmap functions.
 * Re-entrant core libraries.
 * RTMP remoting support.
 * Support for PNG and GIF files in all DefineBitsJpeg tags; support
   for DefineBitsJpeg4 tags, and fix display of some GIFs with
   multiple records.
 * Restore support for mingw32.
 * fb-gnash: allow specifying size and offset of rendering "window".
 * dump-gnash: improved speed, sync and flexibility.
 * JPEG screenshot output.
 * Support for linear RGB interpolation mode in gradient fills.
 * Clipboard setting (GTK only).
 * More flash.geom.Rectangle functions.
 * Improved OpenBSD compatibility.

Gnash 0.8.8
2010/08/17

Improvements since 0.8.7 release are:

 * Renderers are able to be switched at runtime between Cairo,
   OpenGL, and AGG. 
 * Media handlers can be switched at runtime between ffmpeg and
   Gstreamer.
 * Hardware video decoding support via VAAPI for Nvidia, ATI, and
   Intel GPUs. 
 * Android specific support.
 * Gnash now compiles faster due to reduced internal dependencies.
 * Mostly complete Scriptable Plugin support which works with the
   ExternalInterface class.
 * Improved input device handling when using a raw framebuffer.
 * 100% of all YouTube videos should work
 * Support for MovieClip.beginBitmapFill() and gradient spread modes.

Gnash 0.8.7
2010/02/15

Improvements since 0.8.6 release are:

 * Automatic and spontaneous screenshots support in all GUIs (Benjamin Wolsey)
 * Significant memory savings in parsing large XML trees (Benjamin Wolsey)
 * Significant memory savings in functions calls when super is not required
   (Benjamin Wolsey)
 * Enhancements in video streaming [youtube HD button and seeks]
   (Sandro Santilli)
 * Non blocking load of bitmaps, movies, data (Sandro Santilli)
 * Fix use of imported library symbols (Sandro Santilli)
 * AS2 Array handling fixed.  (Benjamin Wolsey)
 * Refactoring to eliminate most static data and get closer to re-entrant VM.
   (Benjamin Wolsey)
 * SWF6 dynamic gradients tested and implemented correctly. 
   (Benjamin Wolsey)
 * TextFormat.getTextExtent implemented. 
   (Benjamin Wolsey)
 * TextField.autoSize implemented for right and center autosizing. 
   (Benjamin Wolsey)

Gnash 0.8.6
2009/09/15

Improvements since 0.8.5 release are:

 * TextSnapshot, Camera, Microphone, TextField and TextFormat implemented.
 * ASSetNative, ASSetNativeAccessor, ASconstructor implemented.
 * New utilities for finding input devices for Camera and Microphone.
 * Fix bitmap smoothing handling in AGG renderer (Bug #25894).
 * Custom Gnash GTK widget.
 * Various improvements to text handling.
 * large-scale refactoring of object and class creation for improved
   AS compatibility.
 * extension of native classes now works in many cases.
 * PyGTK module support.
 * Native and built-in functions properly separated, more native functions
   added.
 * AmigaOS4 support (gui and sound handler).
 * Fixes to XMLSocket crash and event order.
 * Performance improvement for SDL gui (invalidated regions).
 * Corrected timing of various AS events, fixing various SWFs that rely on them.
 * Many minor bug fixes.
 * Better A/V synchronization.

utilities/flvdumper.cpp
// 
//   Copyright (C) 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012
//   Free Software Foundation, Inc
// 
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

#ifdef HAVE_CONFIG_H
#include "gnashconfig.h"
#endif

#include "GnashFileUtilities.h"
#include "GnashSystemNetHeaders.h"

#include <boost/shared_ptr.hpp>
#include <dirent.h>
#include <iostream>
#include <cstdarg>
#include <cstring>
#include <string>
#include <vector>

#ifdef ENABLE_NLS
# include <clocale>
#endif

#include "log.h"
#include "rc.h"
#include "amf.h"
#include "flv.h"
#include "buffer.h"
#include "arg_parser.h"

using namespace cygnal;
using namespace std;
using namespace gnash;

namespace {
gnash::LogFile& dbglogfile = gnash::LogFile::getDefaultInstance();
gnash::RcInitFile& rcfile = gnash::RcInitFile::getDefaultInstance();
}

#ifdef BOOST_NO_EXCEPTIONS
namespace boost
{

	void throw_exception(std::exception const & e)
	{
		std::abort();
	}
}
#endif

static void usage ();

static const char *codec_strs[] = {
    "None",
    "None",
    "H263",
    "Screen",
    "VP6",
    "VP6_Alpha",
    "Screen2",
    "Theora",
    "Dirac",
    "Speex"
};

static const char *format_strs[] = {
    "Uncompressed",
    "ADPCM",
    "MP3",
    "Unknown",
    "Unknown",
    "Nellymoser_8KHZ",
    "Nellymoser",
    // These next are only supported by Gnash
    "Vorbis"
};

static const char *frame_strs[] = {
    "NO_frame",
    "Keyframe",
    "Interframe",
    "Disposable"
};

static const char *size_strs[] = {
    "8Bit",
    "16Bit"
};

static const char *type_strs[] = {
    "Mono",
    "Stereo"
};

static const char *rate_strs[] = {
    "55Khz",
    "11Khz",
    "22Khz",
    "44Khz",
};

int
main(int argc, char *argv[])
{
    bool dump = false;          // dump the FLV data
    bool all = false;		// dump all the tags too
    bool meta = true;		// dump all Meta tags only
    vector<string> infiles;
    
    // Enable native language support, i.e. internationalization
#ifdef ENABLE_NLS
    std::setlocale (LC_ALL, "");
    bindtextdomain (PACKAGE, LOCALEDIR);
    textdomain (PACKAGE);
#endif
    const Arg_parser::Option opts[] =
        {
            { 'h', "help",          Arg_parser::no  },
            { 'v', "verbose",       Arg_parser::no  },
            { 'd', "dump",          Arg_parser::no  },
            { 'a', "all",           Arg_parser::no  },
            { 'm', "meta",           Arg_parser::no  },
        };
    
    Arg_parser parser(argc, argv, opts);
    if( ! parser.error().empty() ) {
        cout << parser.error() << endl;
        exit(EXIT_FAILURE);
    }
    
    for( int i = 0; i < parser.arguments(); ++i ) {
        const int code = parser.code(i);
        try {
            switch( code ) {
              case 'h':
                  usage ();
                  exit(EXIT_SUCCESS);
              case 'v':
                  dbglogfile.setVerbosity();
                  log_debug(_("Verbose output turned on"));
                  break;
              case 'd':
                  dump = true;
                  break;
              case 'a':
                  all = true;
                  break;
              case 'm':
                  meta = true;
                  break;
	      case 0:
		  infiles.push_back(parser.argument(i));
		  break;
	    }
        }
        
        catch (Arg_parser::ArgParserException &e) {
            cerr << _("Error parsing command line options: ") << e.what() << endl;
            cerr << _("This is a Gnash flvdumper bug.") << endl;
        }
    }
    
    if (infiles.empty()) {
        cerr << _("Error: no input file was specified. Exiting.") << endl;
        usage();
        return EXIT_FAILURE;
    }

    // Get the filename from the command line
    string filespec = infiles[0];

    Flv flv; 
    struct stat st;

//    boost::shared_ptr<Flv::flv_header_t> head;
    Flv::previous_size_t   previous = 0;
    boost::shared_ptr<Flv::flv_tag_t> tag;
    
    // Make sure it's an FLV file
    if (stat(filespec.c_str(), &st) == 0) {
	try {
            // Open the binary file
	    ifstream ifs(filespec.c_str(), ios::binary);
	    boost::shared_ptr<cygnal::Buffer> buf(new Buffer);
            // Read just the initial 9 byte header
	    ifs.read(reinterpret_cast<char *>(buf->reference()), sizeof(Flv::flv_header_t));
	    log_debug("header is: %s",  hexify(buf->reference(), 9, false));
	    boost::shared_ptr<Flv::flv_header_t> head = flv.decodeHeader(buf);
	    if (head == 0) {
		log_error("Couldn't decode the header! %s",  hexify(buf->reference(), 9, false));
		exit(EXIT_FAILURE);
	    }
	    if ((head->type & Flv::FLV_VIDEO) && (head->type & Flv::FLV_AUDIO)) {
                cout <<"FLV File type: Video and Audio" << endl;
            } else if (head->type && Flv::FLV_VIDEO) {
		cout << "FLV File type: Video" << endl;
            } else if (head->type && Flv::FLV_AUDIO) {
		cout <<"FLV File type: Audio" << endl;
	    }
	    
	    cout << "FLV Version: " << int(head->version) << " (should always be 1)" << endl;
	    boost::uint32_t headsize = flv.convert24(head->head_size);
	    if (all) {
		cout << "FLV Header size: " << headsize << " (should always be 9)" << endl;
	    }
            // Extract all the Tags
            //size_t total = st.st_size - sizeof(Flv::flv_header_t);
             while (!ifs.eof()) {
		 ifs.read(reinterpret_cast<char *>(&previous), sizeof(Flv::previous_size_t));
		 if (ifs.gcount() != sizeof(Flv::previous_size_t)) {
		     log_error("Couldn't read the entire header");
		 }
		 
		 previous = ntohl(previous);
		 //total -= sizeof(Flv::previous_size_t);
		 if (all) {   
		     cout << "FLV Previous Tag Size was: " << previous << endl;
		 }
		 ifs.read(reinterpret_cast<char *>(buf->reference()), sizeof(Flv::flv_tag_t));
		 if (ifs.gcount() != sizeof(Flv::flv_tag_t)) {
		     log_error("Couldn't read the entire tag");
		 }
		 tag  = flv.decodeTagHeader(buf);
		 if (dump) {
		     flv.dump();
		 }
		 //total -= sizeof(Flv::previous_size_t);
		 size_t bodysize = flv.convert24(tag->bodysize);
		 if (bodysize == 0) {
		     cerr << "FLV Tag size is zero, skipping reading packet body " << bodysize << endl;
		     continue;
		 } else {
		     if (all) {   
			 cout << "FLV Tag size is: " << bodysize +  sizeof(Flv::previous_size_t) << endl;
		     }
		 }
		 buf->resize(bodysize);
		 ifs.read(reinterpret_cast<char *>(buf->reference()), bodysize);
// 		 if (ifs.gcount() != bodysize) {
// 		     log_error("Couldn't read the entire body");
// 		 }
		 //total -= bodysize;
		 switch (tag->type) {
		   case Flv::TAG_AUDIO:
		   {
		       if (all) {
			   cerr << "FLV Tag type is: Audio" << endl;
 			   boost::shared_ptr<Flv::flv_audio_t> data = flv.decodeAudioData(*(buf->reference() + sizeof(Flv::flv_tag_t)));
 			   cout << "\tSound Type is: "   << type_strs[data->type] << endl;
 			   cout << "\tSound Size is: "   << size_strs[data->size] << endl;
 			   cout << "\tSound Rate is: "   << rate_strs[data->rate] << endl;
 			   cout << "\tSound Format is: " << format_strs[data->format] << endl;
		       }
		       break;
		   }
		   case Flv::TAG_VIDEO:
		   {
		       if (all) {
			   cout << "FLV Tag type is: Video" << endl;
 			   boost::shared_ptr<Flv::flv_video_t> data = flv.decodeVideoData(*(buf->reference() + sizeof(Flv::flv_tag_t)));
 			   cout << "\tCodec ID is: "   << codec_strs[data->codecID] << endl;
 			   cout << "\tFrame Type is: " << frame_strs[data->type] << endl;
		       }
		       break;
		   }
		   case Flv::TAG_METADATA:
 		       if (meta || all) {
 			   cout << "FLV Tag type is: MetaData" << endl;
 		       }
		       boost::shared_ptr<cygnal::Element> metadata = flv.decodeMetaData(buf->reference(), bodysize);
		       if (meta && metadata) {
			   metadata->dump();
		       }
		       continue;
		 };		 
             };
        } catch (std::exception& e) {
	    log_error("Reading  %s: %s", filespec, e.what());
	    return false;
	}
    }
}

/// \brief  Display the command line arguments
static void
usage ()
{
    cerr << _("This program dumps the internal data of an FLV video file")
         << endl;
    cerr << _("Usage: flvdumper [-h] [-m] [-a] filename") << endl;
    cerr << _("-h\tHelp") << endl;
    cerr << _("-m\tPrint only Meta tags (default)") << endl;
    cerr << _("-a\tPrint all tags.") << endl;
    exit (-1);
}

// Local Variables:
// mode: C++
// indent-tabs-mode: t
// End:
